vecstore: implement partition manipulation functions

mw5h opened the pull request:
Stub out the persistent storage vecstore implementation. Implement the methods for partition creation, retrieval and deletion. Add a simple test to ensure that these functions work properly.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137319)
<!-- Reviewable:end -->

DrewKimball reviewed (commented):
Nice! Mostly just nits, and one concern about a possible bug.

> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-:-ODu8Tay0yjF5ZwcnQhn:biubcg1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/vecindex/vecstore/persistent_store.go` line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODszp28DBgjPzH140KA:-ODszp28DBgjPzH140KB:bpkazkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_store.go#L17)):*
> ```Go
> 	db        *kv.DB // Needed for index maintenance functions
> 	quantizer quantize.Quantizer
> 	prefix    []byte
> ```

[nit] Can we make this `roachpb.Key`?
___
*[`pkg/sql/vecindex/vecstore/persistent_store.go` line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODsxODIEkSP7v245s-x:-ODsxODIEkSP7v245s-y:b4yyou1) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_store.go#L19)):*
> ```Go
> 	prefix    []byte
> 	dims      int
> 	seed      int64
> ```

[nit] doesn't look like `dims` or `seed` are used here, and it seems like `seed` should never be needed. `dims` could maybe be accessed through `quantizer`, though I'm fine with keeping it here if we do end up needing it.
___
*[`pkg/sql/vecindex/vecstore/persistent_store_test.go` line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODt1mYt7GSZC6x8MOQW:-ODt1mYt7GSZC6x8MOQX:b-6sfqse) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_store_test.go#L40)):*
> ```Go
> 	store := NewPersistentStore(kvDB, quantizer, prefix, 2, 0)
> 
> 	t.Run("insert a root partition into the store and read it back", func(t *testing.T) {
> ```

Could we use this to verify the partition is correct? https://github.com/cockroachdb/cockroach/blob/e8ee6f574ddf1fce1a4cb53f392c5a9baf633b76/pkg/sql/vecindex/vecstore/encoding_test.go#L131
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODu63vOC21MgRA2meqj:-ODu63vOC21MgRA2meqk:basbv2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L76)):*
> ```Go
> 	// partition metadata but is part of the data that all implementations
> 	// QuantizedVectorSet need.
> 	vectorSet := psTxn.store.quantizer.Quantize(ctx, &vector.Set{})
> ```

[nit] It seems strange to quantize the empty set - can we instead have `var vectorSet QuantizedVectorSet` and then initialize it with `MakeRaBitQCodeSet` and `&UnQuantizedVectorSet{Centroid: centroid}` respectively in the switch?

Alternatively, maybe it's time to add some methods to either the `QuantizedVectorSet` interface, or `Partition`, to avoid having to switch on the quantization type all the time.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODu2vh09qOemRrEf2LB:-ODu2vh09qOemRrEf2LC:b-26jfc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L142)):*
> ```Go
> 
> 	// Store the length of the partition's metadata key so that we can truncate back
> 	// to that length instead of building a new key from scratch.
> ```

[nit] This comment looks stale.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODsqRBD9tWqYPuKyiRH:-ODsqRBD9tWqYPuKyiRI:b-1z4f5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L166)):*
> ```Go
> 	// batch? Could this be batched with the insertPartition? Docs say operations
> 	// happen out of order, but maybe there's something to keep a single batch from
> 	// self-interfering?
> ```

Which docs are you referring to? I'd expect write batches to evaluate requests in order. Separate batches seems fine for now, though.
https://github.com/cockroachdb/cockroach/blob/e8ee6f574ddf1fce1a4cb53f392c5a9baf633b76/pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator.go#L21-L27
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODu4tGo1Qh3pO2UUhd9:-ODu4tGo1Qh3pO2UUhdA:b90c5pc) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L180)):*
> ```Go
> 	return partitionID, psTxn.insertPartition(ctx, partitionID, partition,
> 		func(b *kv.Batch, k roachpb.Key, v []byte) {
> 			b.CPutAllowingIfNotExists(k, v, nil /* expValue */)
> ```

[nit] CPut has the same behavior for nil `expValue`, right?
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 215 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODu5c3p7fKmqUeg0-ao:-ODu5c3q2IPR7TwHCzGf:b-yj7efb) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L215)):*
> ```Go
> 	partitionCounts []int,
> ) (Level, error) {
> 	panic("SearchPartitions() unimplemented")
> ```

I think we could implement this relatively easily by calling `GetPartition`
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODt-fqG1oOQYQo9-i1Q:-ODt-fqG1oOQYQo9-i1R:b702ond) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L226)):*
> ```Go
> // the next partition's metadata.
> func (psTxn *PersistentStoreTxn) encodePartitionKey(partitionKey PartitionKey) roachpb.Key {
> 	psTxn.keyBuffer = psTxn.keyBuffer[:psTxn.keyPrefixLen]
> ```

I'm not certain it's safe to use a buffer like this. I think it's possible for the KV layer to hold on to keys for the duration of the txn (to be used for things like read refreshes and async consensus for writes?). I ran into a problem like that while working on encoding the keys for vectorized index joins a while ago.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 227 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODu8Co-7RONWBl8bSnj:-ODu8Co-7RONWBl8bSnk:bqfg672) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L227)):*
> ```Go
> func (psTxn *PersistentStoreTxn) encodePartitionKey(partitionKey PartitionKey) roachpb.Key {
> 	psTxn.keyBuffer = psTxn.keyBuffer[:psTxn.keyPrefixLen]
> 	psTxn.keyBuffer = encoding.EncodeUvarintAscending(psTxn.keyBuffer, uint64(partitionKey))
> ```

[nit] Let's add an `EncodePartitionKey` helper function as well.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/vecindex/vecstore/persistent_store.go` line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODsxODIEkSP7v245s-x:-ODvQETA7agJ9fHY26VA:b-cv7hej) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_store.go#L19)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] doesn't look like `dims` or `seed` are used here, and it seems like `seed` should never be needed. `dims` could maybe be accessed through `quantizer`, though I'm fine with keeping it here if we do end up needing it.
</blockquote></details>

I think we should drop the `dims`, since there's no situation where it can ever be different than `Quantizer.GetOriginalDims` / `Quantizer.GetRandomDims`. Having it separate could cause confusion/bugs over which it is and whether it can be different.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODu63vOC21MgRA2meqj:-ODvTNmT8izpNo1vhIpo:b-iiz5yo) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L76)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] It seems strange to quantize the empty set - can we instead have `var vectorSet QuantizedVectorSet` and then initialize it with `MakeRaBitQCodeSet` and `&UnQuantizedVectorSet{Centroid: centroid}` respectively in the switch?

Alternatively, maybe it's time to add some methods to either the `QuantizedVectorSet` interface, or `Partition`, to avoid having to switch on the quantization type all the time.
</blockquote></details>

I think we have a few options to discover the right quantization type:
1. We could store it in the partition metadata.
2. We could add methods to `Quantizer` or `QuantizedSet` that decode/encode.
3. We could switch on the type of `psTxn.store.quantizer` or the set it produces like we do here.

The concern with #1 is that there might be cases where we'd have to fetch the metadata just to know how to decode the partition data - I'm not sure it's worth the risk.

The concern with #2 is that it seems like encoding/decoding is a Store concern, not a Quantizer concern. Maybe the in-memory store wants to encode/decode using protobuf and the persistent store wants to use a SQL-like encoding, which seems like the right separation of concerns.

For #3, I think we should probably switch on the type `psTxn.store.quantizer` rather than creating the empty set. That way, we can preallocate the buffers in `UnQuantizedSet`/`RaBitQuantizedSet` so that they don't need to be resized by the decode methods. We could add a helper method to `UnQuantizedVectorSet` and `RaBitQuantizedVectorSet` to do this, e.g. `NewEmptyRaBitQuantizedVectorSet(capacity int)`.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODu4tGo1Qh3pO2UUhd9:-ODvc30q5lzkZ6r_zqx6:b-pbc6ft) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L180)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] CPut has the same behavior for nil `expValue`, right?
</blockquote></details>

In both cases, we're expecting there to be no existing rows (b/c we deleted the root partition and b/c we generated a unique key for other partitions). Is `expValue` intended to assert that there really is no existing partition in this case? I'd be a bit concerned that `CPut` is quite a bit more expensive than `Put` - maybe worth asking the KV team in case it's not worth the cost of the assertion. Regardless, I'd expect to use `Put` in both cases or `CPut` in both cases, since that would simplify `psTxn.insertPartition` to not require a higher-order function.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODt-fqG1oOQYQo9-i1Q:-ODvd6452jsJjSAa6cHg:b-b1delr) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L226)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I'm not certain it's safe to use a buffer like this. I think it's possible for the KV layer to hold on to keys for the duration of the txn (to be used for things like read refreshes and async consensus for writes?). I ran into a problem like that while working on encoding the keys for vectorized index joins a while ago.
</blockquote></details>

If you do switch to always allocate here, you should add a separate helper function that does reuse `keyBuffer` for the `PrefixEnd` case. That function allocates a new buffer anyways, so no reason to allocate twice.
___
*[`pkg/sql/vecindex/vecstore/encoding.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODvOxKA-307Hkhkt850:-ODvOxKA-307Hkhkt851:b-es9ml5) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/encoding.go#L55)):*
> ```Go
> 		return append(appendTo, key.PrimaryKey...)
> 	}
> 	return encoding.EncodeUvarintAscending(appendTo, uint64(key.PartitionKey))
> ```

How come you switched to the varint encoding? I seem to remember @DrewKimball believing that the fixed length encoding was preferable for this use case, maybe he can describe his reasoning.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODvSBSI8peUZaGQvwc4:-ODvSBSI8peUZaGQvwc5:bkbvh2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L58)):*
> ```Go
> 	endKey := startKey.PrefixEnd()
> 
> 	b.Scan(startKey, endKey)
> ```

I think we'll need to lock the keyspan that we read here, e.g. using equivalent of `SELECT FOR UPDATE`. `GetPartition` is only used by split/merge, and we want any concurrent insert/delete ops on this partition to block until commit.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODvV_Ol4q_hx2CImbh9:-ODvV_Om1PaR1LErpDEk:b-v3p2un) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L76)):*
> ```Go
> 	// partition metadata but is part of the data that all implementations
> 	// QuantizedVectorSet need.
> 	vectorSet := psTxn.store.quantizer.Quantize(ctx, &vector.Set{})
> ```

For the root partition, we always want to use `UnQuantizer`. We can check for that case using `partitionKey == RootKey`. All other partitions should use the deserialization rules for `psTxn.store.quantizer`.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODv_8IB8wWSGK5Y5jSU:-ODv_8IB8wWSGK5Y5jSV:b-cw8l2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L177)):*
> ```Go
> 	ctx context.Context, partition *Partition,
> ) (PartitionKey, error) {
> 	partitionID := PartitionKey(builtins.GenerateUniqueInt(builtins.ProcessUniqueID(psTxn.store.db.Context().NodeID.SQLInstanceID())))
> ```

NIT: line longer than 100 chars, also would be good to break up expression for more clarity.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/vecindex/vecstore/encoding.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODvOxKA-307Hkhkt850:-ODwDR796E1ooPHjV4d2:b-3ofgv6) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/encoding.go#L55)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

How come you switched to the varint encoding? I seem to remember @DrewKimball believing that the fixed length encoding was preferable for this use case, maybe he can describe his reasoning.
</blockquote></details>

It seems fine in this case because this is the last thing encoded in the key, so we can know the offset of each encoded element. The PK encoding for leaf partitions will be variable length, anyway.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODt-fqG1oOQYQo9-i1Q:-ODwEk7q6ylh7GC3R8gI:b-vgp98u) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L226)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

If you do switch to always allocate here, you should add a separate helper function that does reuse `keyBuffer` for the `PrefixEnd` case. That function allocates a new buffer anyways, so no reason to allocate twice.
</blockquote></details>

What do you mean by this? Where exactly would the buffer be used?


<!-- Sent from Reviewable.io -->

mw5h committed:
vecstore: implement partition manipulation functions

Add persistent storage to vecstore. Implement the methods for partition
creation, retrieval and deletion. Add a simple test to ensure these
methods are functioning.

Epic: CRDB-42943

mw5h reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @DrewKimball)

___
*[`pkg/sql/vecindex/vecstore/encoding.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODvOxKA-307Hkhkt850:-ODwCUME0t5ocJruG8hs:b-ka9pyw) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/encoding.go#L55)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

How come you switched to the varint encoding? I seem to remember @DrewKimball believing that the fixed length encoding was preferable for this use case, maybe he can describe his reasoning.
</blockquote></details>

I did this while debugging an unrelated issue. The fixed with encoding was being displayed as a series of NULLs and a single byte of data. It made me wonder if the KV layer didn't like the input, since the rest of the key prefix is built using varint encoding. My problem was elsewhere, but I left this change in because it seemed more in line with what was happening elsewhere.
___
*[`pkg/sql/vecindex/vecstore/persistent_store.go` line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODszp28DBgjPzH140KA:-ODx4mdS2NuCCh900q2p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_store.go#L17)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Can we make this `roachpb.Key`?
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/vecstore/persistent_store.go` line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODsxODIEkSP7v245s-x:-ODwD3KN9D9Hph-wQJie:btd51a6) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_store.go#L19)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think we should drop the `dims`, since there's no situation where it can ever be different than `Quantizer.GetOriginalDims` / `Quantizer.GetRandomDims`. Having it separate could cause confusion/bugs over which it is and whether it can be different.
</blockquote></details>

I agree.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODvSBSI8peUZaGQvwc4:-ODwDF5A3eDpnDGeI74l:bi78xuu) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L58)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think we'll need to lock the keyspan that we read here, e.g. using equivalent of `SELECT FOR UPDATE`. `GetPartition` is only used by split/merge, and we want any concurrent insert/delete ops on this partition to block until commit.
</blockquote></details>

Sure, that makes sense.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODu63vOC21MgRA2meqj:-ODwDL0IEi0BL46D6mbe:bo2krpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L76)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think we have a few options to discover the right quantization type:
1. We could store it in the partition metadata.
2. We could add methods to `Quantizer` or `QuantizedSet` that decode/encode.
3. We could switch on the type of `psTxn.store.quantizer` or the set it produces like we do here.

The concern with #1 is that there might be cases where we'd have to fetch the metadata just to know how to decode the partition data - I'm not sure it's worth the risk.

The concern with #2 is that it seems like encoding/decoding is a Store concern, not a Quantizer concern. Maybe the in-memory store wants to encode/decode using protobuf and the persistent store wants to use a SQL-like encoding, which seems like the right separation of concerns.

For #3, I think we should probably switch on the type `psTxn.store.quantizer` rather than creating the empty set. That way, we can preallocate the buffers in `UnQuantizedSet`/`RaBitQuantizedSet` so that they don't need to be resized by the decode methods. We could add a helper method to `UnQuantizedVectorSet` and `RaBitQuantizedVectorSet` to do this, e.g. `NewEmptyRaBitQuantizedVectorSet(capacity int)`.
</blockquote></details>

The next version adds a method to the Quantizer interface to produce a pre-allocated but empty QuantizedSet. I still switch on the quantized set type, but this feels quite a bit cleaner.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODvV_Ol4q_hx2CImbh9:-ODwDW0GAe1z0i0CCuAl:b-10xlom) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L76)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

For the root partition, we always want to use `UnQuantizer`. We can check for that case using `partitionKey == RootKey`. All other partitions should use the deserialization rules for `psTxn.store.quantizer`.
</blockquote></details>

Oh, that's important!
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODu2vh09qOemRrEf2LB:-OE0xkxr2B7mzcZs7qBf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L142)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] This comment looks stale.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODsqRBD9tWqYPuKyiRH:-OE0xof5AfvnwDokz1_N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L166)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Which docs are you referring to? I'd expect write batches to evaluate requests in order. Separate batches seems fine for now, though.
https://github.com/cockroachdb/cockroach/blob/e8ee6f574ddf1fce1a4cb53f392c5a9baf633b76/pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator.go#L21-L27
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODv_8IB8wWSGK5Y5jSU:-OE0yG9tCR4in2f1fjet:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L177)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: line longer than 100 chars, also would be good to break up expression for more clarity.
</blockquote></details>

Fixed.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODu4tGo1Qh3pO2UUhd9:-ODwDfYK4jNM_nyUlbxc:b-mch12r) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L180)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

In both cases, we're expecting there to be no existing rows (b/c we deleted the root partition and b/c we generated a unique key for other partitions). Is `expValue` intended to assert that there really is no existing partition in this case? I'd be a bit concerned that `CPut` is quite a bit more expensive than `Put` - maybe worth asking the KV team in case it's not worth the cost of the assertion. Regardless, I'd expect to use `Put` in both cases or `CPut` in both cases, since that would simplify `psTxn.insertPartition` to not require a higher-order function.
</blockquote></details>

The intention was to ensure that we're not overwriting data, yes. I can lose this in favor of Put().
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODt-fqG1oOQYQo9-i1Q:-ODwEEyqAmnZ6YGpwcZh:b55m64w) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L226)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

If you do switch to always allocate here, you should add a separate helper function that does reuse `keyBuffer` for the `PrefixEnd` case. That function allocates a new buffer anyways, so no reason to allocate twice.
</blockquote></details>

Okay, thanks for the info.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 227 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODu8Co-7RONWBl8bSnj:-OE1-C_n5eoloyW7EQoa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L227)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Let's add an `EncodePartitionKey` helper function as well.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/vecstore/persistent_store_test.go` line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODt1mYt7GSZC6x8MOQW:-ODwEW_1AKAkyBimxsXk:b-g2i9kc) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_store_test.go#L40)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could we use this to verify the partition is correct? https://github.com/cockroachdb/cockroach/blob/e8ee6f574ddf1fce1a4cb53f392c5a9baf633b76/pkg/sql/vecindex/vecstore/encoding_test.go#L131
</blockquote></details>

I bet we could.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/vecindex/quantize/unquantizer.go` line 80 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEFF7es5w8-j6XyqZzF:-OEFF7et-XxByLKnMGOD:bpx27ru) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1cff40e3d36ae5b0f262c2411bcf7bad02e294d/pkg/sql/vecindex/quantize/unquantizer.go#L80)):*
> ```Go
> // NewQuantizedVectorSet implements the Quantizer interface
> func (q *unQuantizer) NewQuantizedVectorSet(size int) QuantizedVectorSet {
> 	dataBuffer := make([]float32, 0, size*q.GetOriginalDims())
> ```

Same here, for consistency, even though it's the same value.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 20 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEFFa4cCHHG0oO9LZ88:-OEFFa4cCHHG0oO9LZ89:blzwllq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1cff40e3d36ae5b0f262c2411bcf7bad02e294d/pkg/sql/vecindex/vecstore/persistent_txn.go#L20)):*
> ```Go
> )
> 
> // The PersistentStoreTxn provides a context to make transactional changes to a
> ```

NIT: Linter might not like that comment does not begin with name of the struct - by convention, should drop "The".
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 33 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEFFj_59gODsdYieGPT:-OEFFj_59gODsdYieGPU:bedmb5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1cff40e3d36ae5b0f262c2411bcf7bad02e294d/pkg/sql/vecindex/vecstore/persistent_txn.go#L33)):*
> ```Go
> var _ Txn = (*PersistentStoreTxn)(nil)
> 
> func NewPersistentStoreTxn(store *PersistentStore, kv *kv.Txn) *PersistentStoreTxn {
> ```

NIT: comment for exported func.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 38 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEFG1qz6-DnzEuBU0la:-OEFG1qz6-DnzEuBU0lb:be6yrs9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1cff40e3d36ae5b0f262c2411bcf7bad02e294d/pkg/sql/vecindex/vecstore/persistent_txn.go#L38)):*
> ```Go
> 		store: store,
> 	}
> 	// TODO (mw5h): This doesn't take into account session variables that control
> ```

You should maybe check with the KV team on whether we need `GuaranteedDurability` for non-Serializable isolation levels for this scenario.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 72 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEFI-vZCQG60ME29Lcn:-OEFI-vZCQG60ME29Lco:b7o661a) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1cff40e3d36ae5b0f262c2411bcf7bad02e294d/pkg/sql/vecindex/vecstore/persistent_txn.go#L72)):*
> ```Go
> 		return nil, ErrPartitionNotFound
> 	}
> 	// Partition metadata is stored in /Prefix/PartitionID, with vector data following in /Prefix/PartitionID/ChildKey
> ```

NIT: break comment lines at 80 chars, above as well.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 233 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEFHfLbDFeCShRNvkPY:-OEFHfLbDFeCShRNvkPZ:bcpe6g0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1cff40e3d36ae5b0f262c2411bcf7bad02e294d/pkg/sql/vecindex/vecstore/persistent_txn.go#L233)):*
> ```Go
> // the next partition's metadata.
> func (psTxn *PersistentStoreTxn) encodePartitionKey(partitionKey PartitionKey) roachpb.Key {
> 	keyBuffer := append(roachpb.Key{}, psTxn.store.prefix...)
> ```

This looks like it might have to reallocate the underlying array buffer in `EncodePartitionKey` - you should preallocate enough bytes to prevent that.
___
*[`pkg/sql/vecindex/vecstore/persistent_store.go` line 24 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEFFJbRAS6Y70GWE0FO:-OEFFJbRAS6Y70GWE0FP:bc3hzte) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1cff40e3d36ae5b0f262c2411bcf7bad02e294d/pkg/sql/vecindex/vecstore/persistent_store.go#L24)):*
> ```Go
> var _ Store = (*PersistentStore)(nil)
> 
> func NewPersistentStore(
> ```

NIT: should have comments on exported funcs.
___
*[`pkg/sql/vecindex/vecstore/persistent_store.go` line 39 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEFFTBh2eJtS7nLpvH5:-OEFFTBh2eJtS7nLpvH6:bexk9yi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1cff40e3d36ae5b0f262c2411bcf7bad02e294d/pkg/sql/vecindex/vecstore/persistent_store.go#L39)):*
> ```Go
> }
> 
> // Begin() is part of the vecstore.Store interface. Begin() creates a new KV
> ```

NIT: style-wise, we don't add () parens in comments.
___
*[`pkg/sql/vecindex/quantize/rabitq.go` line 147 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEFEcAGEV3o2I7Kd8RB:-OEFEcAGEV3o2I7Kd8RC:b-xzlq4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1cff40e3d36ae5b0f262c2411bcf7bad02e294d/pkg/sql/vecindex/quantize/rabitq.go#L147)):*
> ```Go
> // NewQuantizedVectorSet implements the Quantizer interface
> func (q *raBitQuantizer) NewQuantizedVectorSet(size int) QuantizedVectorSet {
> 	dataBuffer := make([]uint64, 0, size*RaBitQCodeSetWidth(q.GetOriginalDims()))
> ```

This should use `GetRandomDims`, because the vector set is populated after the randomization step.


<!-- Sent from Reviewable.io -->

mw5h reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @DrewKimball)

___
*[`pkg/sql/vecindex/quantize/rabitq.go` line 147 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEFEcAGEV3o2I7Kd8RB:-OEQGQc06gR1mLq6JO6R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1cff40e3d36ae5b0f262c2411bcf7bad02e294d/pkg/sql/vecindex/quantize/rabitq.go#L147)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This should use `GetRandomDims`, because the vector set is populated after the randomization step.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/quantize/unquantizer.go` line 80 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEFF7es5w8-j6XyqZzF:-OEQGR-0BtDBK1natjGj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1cff40e3d36ae5b0f262c2411bcf7bad02e294d/pkg/sql/vecindex/quantize/unquantizer.go#L80)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Same here, for consistency, even though it's the same value.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/vecstore/encoding.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODvOxKA-307Hkhkt850:-OEQGU_sBeo8jBnxYd69:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/encoding.go#L55)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

I did this while debugging an unrelated issue. The fixed with encoding was being displayed as a series of NULLs and a single byte of data. It made me wonder if the KV layer didn't like the input, since the rest of the key prefix is built using varint encoding. My problem was elsewhere, but I left this change in because it seemed more in line with what was happening elsewhere.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 215 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODu5c3p7fKmqUeg0-ao:-OEQI4Hq8O5IAKaf7VmR:b-cni1ok) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L215)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I think we could implement this relatively easily by calling `GetPartition`
</blockquote></details>

GetPartition takes a lock, so no. I'll factor the partition decoding out so that it can be reused.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-ODt-fqG1oOQYQo9-i1Q:-OEQI0TtDNglAkx112D3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da2a5c5228f067b673a8d3a82ba2a4d7a07e78e6/pkg/sql/vecindex/vecstore/persistent_txn.go#L226)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

Okay, thanks for the info.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 38 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEFG1qz6-DnzEuBU0la:-OEQAv45ACtb83nEjBcC:b-6gxjv4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1cff40e3d36ae5b0f262c2411bcf7bad02e294d/pkg/sql/vecindex/vecstore/persistent_txn.go#L38)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

You should maybe check with the KV team on whether we need `GuaranteedDurability` for non-Serializable isolation levels for this scenario.
</blockquote></details>

This is encoded in optbuilder, but the clean way of accessing it is a method that we don't have access to. 
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 233 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEFHfLbDFeCShRNvkPY:-OEQDQ615V6SNBV0xMMu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1cff40e3d36ae5b0f262c2411bcf7bad02e294d/pkg/sql/vecindex/vecstore/persistent_txn.go#L233)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This looks like it might have to reallocate the underlying array buffer in `EncodePartitionKey` - you should preallocate enough bytes to prevent that.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 5 files at r7, 8 of 8 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-:-OEQoZro62p3H1lA0XWX:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 77 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEQlgkc9WtyRF47XH-0:-OEQlgkd6A6ZhJOOn5oG:bebg6wu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c10b1f085e61e7baa6be9a88ddffd2b1c6ba5b5/pkg/sql/vecindex/vecstore/persistent_txn.go#L77)):*
> ```Go
> 	}
> 	// Partition metadata is stored in /Prefix/PartitionID, with vector data
> 	// following in /Prefix/PartitionID/ChildKey
> ```

nit: comment needs a period
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 92 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OEQltUF87RyiWTeAEFM:-OEQltUF87RyiWTeAEFN:bpbdlon) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c10b1f085e61e7baa6be9a88ddffd2b1c6ba5b5/pkg/sql/vecindex/vecstore/persistent_txn.go#L92)):*
> ```Go
> 	// Build a decoder function for deserializing encoded vectors. We also set the
> 	// centroid of the empty vector set here because the centroid is stored in the
> 	// partition metadata but is part of the data that all implementations
> ```

nit: should be: `all implementations of`


<!-- Sent from Reviewable.io -->

mw5h committed:
vecstore: implement AddTo/RemoveFrom/Search functions

Add implementations of AddToPartition, RemoveFromPartition and
SearchPartitions. AddToPartition and RemoveFromPartition turn into
simple KV operations since each encoded vector is written as a single KV
value. SearchPartitions reuses a lot of the code for GetPartition, but
without the locking.

Epic: CRDB-42943

mw5h committed:
vecstore: move vector store implementation tests into a common test

Previously, the in-memory and persistent vector stores maintained their
own disjoint tests. This patch moves those tests that test common
functionality into a single joint test that is referenced from both
implementation unit tests.

Epic: CRDB-42943

DrewKimball reviewed (commented):
> Reviewed 6 of 6 files at r9, 2 of 2 files at r10, 4 of 4 files at r11, 1 of 1 files at r12, 6 of 6 files at r13, 2 of 2 files at r14, 4 of 4 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/vecindex/vecstore/store_test.go` line 17 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFOEuVf1b3InVKoHlla:-OFOEuVf1b3InVKoHllb:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aad54a088728cbdadf9a29def703bdda000158d/pkg/sql/vecindex/vecstore/store_test.go#L17)):*
> ```Go
> )
> 
> func commonStoreTests(
> ```

Nice!
___
*[`pkg/sql/vecindex/quantize/quantizer.go` line 120 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFO1cvT9-se3fZmoiEv:-OFO1cvT9-se3fZmoiEw:bq5me5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1f431f952c7b0b52e670d6dd9252c4cb1d5bfc3/pkg/sql/vecindex/quantize/quantizer.go#L120)):*
> ```Go
> 	// Clear removes all the elements of the vector set so that it may be reused. The
> 	// new centroid is copied over the existing centroid.
> 	Clear(centroid vector.T)
> ```

[nit] Given that this method sets a new centroid and prepares the set for reuse, would `Reset()` be a better name?
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 236 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFO50vo7iSkaaybeZAI:-OFO50vp7U-1a36pG2K9:b-w4mmy8) ([raw file](https://github.com/cockroachdb/cockroach/blob/558ba4c35c2bc62c3b8bb2ebc8d1023b14b8e4d6/pkg/sql/vecindex/vecstore/persistent_txn.go#L236)):*
> ```Go
> // existing metadata, but existing vectors will not be deleted. Vectors in the
> // new partition will overwrite existing vectors if child keys collide, but
> // otherwise the resulting partition will be a union of the two partitions.
> ```

Should we consider it a bug if a caller is trying to overwrite a partition without having first deleted it?
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 302 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFO6ToO3KUVd9PTRmr-:-OFO6ToO3KUVd9PTRmr0:be7ecvd) ([raw file](https://github.com/cockroachdb/cockroach/blob/43e16d3b596ea568be9e6c0a9737ed0de1d42112/pkg/sql/vecindex/vecstore/persistent_txn.go#L302)):*
> ```Go
> 	ctx context.Context, partitionKey PartitionKey, vector vector.T, childKey ChildKey,
> ) (int, error) {
> 	// TODO(mw5h): Add to an existing batch instead of starting a new one.
> ```

This concern might go away when buffered writes are implemented.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 320 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFO7gKR4Cd6p5ZgcMvZ:-OFO7gKR4Cd6p5ZgcMv_:bgkdb0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/43e16d3b596ea568be9e6c0a9737ed0de1d42112/pkg/sql/vecindex/vecstore/persistent_txn.go#L320)):*
> ```Go
> 	}
> 
> 	entryKey := make([]byte, len(metadataKey))
> ```

[nit] Since we have to reallocate on the append anyway, we could just cap the `metadataKey` and use it directly:
```
// Cap the metadata key so that the append allocates a new slice for the child key.
prefix := metadataKey[:len(metadataKey):len(metadataKey)]
entryKey := EncodeChildKey(prefix, childKey)
```


<!-- Sent from Reviewable.io -->

mw5h reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @DrewKimball)

___
*[`pkg/sql/vecindex/quantize/quantizer.go` line 120 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFO1cvT9-se3fZmoiEv:-OFP1Yn9B_uh2CQkpV7d:bon699g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1f431f952c7b0b52e670d6dd9252c4cb1d5bfc3/pkg/sql/vecindex/quantize/quantizer.go#L120)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Given that this method sets a new centroid and prepares the set for reuse, would `Reset()` be a better name?
</blockquote></details>

Sadly, the protobuf generated code already defines Reset(). Not sure we want to mess with that. I do agree that Clear() leaves something to be desired -- I just grabbed that because there are some Clear methods hanging about that do similar things. The 'centroid' argument is suboptimal.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 236 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFO50vo7iSkaaybeZAI:-OFP2STK6ashynD5bqey:bk4fbvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/558ba4c35c2bc62c3b8bb2ebc8d1023b14b8e4d6/pkg/sql/vecindex/vecstore/persistent_txn.go#L236)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Should we consider it a bug if a caller is trying to overwrite a partition without having first deleted it?
</blockquote></details>

To do that, we'd have to go back to using CPut for writing the partition's metadata. Andy's comments made it sound like this was not free. Are we willing to spend cycles on this?
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 320 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFO7gKR4Cd6p5ZgcMvZ:-OFP2lvd5deY0AAkEyFM:b-5crhuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/43e16d3b596ea568be9e6c0a9737ed0de1d42112/pkg/sql/vecindex/vecstore/persistent_txn.go#L320)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Since we have to reallocate on the append anyway, we could just cap the `metadataKey` and use it directly:
```
// Cap the metadata key so that the append allocates a new slice for the child key.
prefix := metadataKey[:len(metadataKey):len(metadataKey)]
entryKey := EncodeChildKey(prefix, childKey)
```
</blockquote></details>

Done. Thanks for that!


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/vecindex/quantize/quantizer.go` line 120 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFO1cvT9-se3fZmoiEv:-OFQRh3bC9XZyhw1al00:bu7u38g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1f431f952c7b0b52e670d6dd9252c4cb1d5bfc3/pkg/sql/vecindex/quantize/quantizer.go#L120)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

Sadly, the protobuf generated code already defines Reset(). Not sure we want to mess with that. I do agree that Clear() leaves something to be desired -- I just grabbed that because there are some Clear methods hanging about that do similar things. The 'centroid' argument is suboptimal.
</blockquote></details>

I see, too bad.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 236 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFO50vo7iSkaaybeZAI:-OFQRBx04d311OB9E2U4:b-gmayru) ([raw file](https://github.com/cockroachdb/cockroach/blob/558ba4c35c2bc62c3b8bb2ebc8d1023b14b8e4d6/pkg/sql/vecindex/vecstore/persistent_txn.go#L236)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

To do that, we'd have to go back to using CPut for writing the partition's metadata. Andy's comments made it sound like this was not free. Are we willing to spend cycles on this?
</blockquote></details>

I wouldn't think CPut overhead should matter for partition creation, but if it's a concern we could just make it a documented requirement of the function without actually validating it during execution. It's minor though, so I'm fine with whatever you end up deciding.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/vecindex/vecstore/persistent_store.go` line 16 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFwN1rH0ekUwA0vZfoT:-OFwN1rH0ekUwA0vZfoU:b-rsytwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/0faf35ff734b74bf34703442607a91ef294f546b/pkg/sql/vecindex/vecstore/persistent_store.go#L16)):*
> ```Go
> )
> 
> type PersistentStore struct {
> ```

NIT: Add comment for exported type.xN
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 40 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFwNZH1-zSAWIsLXIXL:-OFwNZH1-zSAWIsLXIXM:bufucq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0faf35ff734b74bf34703442607a91ef294f546b/pkg/sql/vecindex/vecstore/persistent_txn.go#L40)):*
> ```Go
> 
> // PersistentStoreCodec abstracts quantizer-specific aspects of reading and writing vector sets.
> type PersistentStoreCodec struct {
> ```

Does this need to be exported?
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 43 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFwNdKZAL8a6copWoBp:-OFwNdKZAL8a6copWoBq:bafw67a) ([raw file](https://github.com/cockroachdb/cockroach/blob/0faf35ff734b74bf34703442607a91ef294f546b/pkg/sql/vecindex/vecstore/persistent_txn.go#L43)):*
> ```Go
> 	quantizer quantize.Quantizer
> 
> 	encoder func(quantize.QuantizedVectorSet, []byte, int) ([]byte, error)
> ```

Why have functors like this, rather than using an interface? It creates extra objects to do it this way, as well as being less idiomatic.

In fact, I'm not even sure an interface is worth it, given the simplicity of what's going on here. You could embed the codec struct implementation directly in `PersistentStoreTxn` so that you don't create any extra objects - `getCodecForPartitionKey` would then just return a pointer to one or the other. The implementation would just do the switch on the quantizer type each time it's called rather than having to create separate objects for dynamic dispatch.


<!-- Sent from Reviewable.io -->

mw5h reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @DrewKimball)

___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 302 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFO6ToO3KUVd9PTRmr-:-OFx8Y3b3jD-v4bxLmCF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/43e16d3b596ea568be9e6c0a9737ed0de1d42112/pkg/sql/vecindex/vecstore/persistent_txn.go#L302)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

This concern might go away when buffered writes are implemented.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 40 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFwNZH1-zSAWIsLXIXL:-OFx8gwg0mR5IPfrUKQK:b-7b6x4z) ([raw file](https://github.com/cockroachdb/cockroach/blob/0faf35ff734b74bf34703442607a91ef294f546b/pkg/sql/vecindex/vecstore/persistent_txn.go#L40)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Does this need to be exported?
</blockquote></details>

No, it doesn't. Good catch.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 43 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFwNdKZAL8a6copWoBp:-OFx8kSV6Vrag2vtXllZ:b-trqnz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0faf35ff734b74bf34703442607a91ef294f546b/pkg/sql/vecindex/vecstore/persistent_txn.go#L43)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Why have functors like this, rather than using an interface? It creates extra objects to do it this way, as well as being less idiomatic.

In fact, I'm not even sure an interface is worth it, given the simplicity of what's going on here. You could embed the codec struct implementation directly in `PersistentStoreTxn` so that you don't create any extra objects - `getCodecForPartitionKey` would then just return a pointer to one or the other. The implementation would just do the switch on the quantizer type each time it's called rather than having to create separate objects for dynamic dispatch.
</blockquote></details>

Yeah, I started with an interface implementation and it just created a lot of boilerplate for very little gain. From there I went to functors because that's the C thing to do. I'll take another shot at it.


<!-- Sent from Reviewable.io -->

mw5h commented:
This is ready for review (as if we hadn't already been doing that). The GetFullVectors changes are big enough that I'll post it as a separate PR.

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 6 files at r20, 12 of 12 files at r23, 8 of 8 files at r24, 10 of 10 files at r25, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-:-OGI0Hv5A-UiGWOsgUEy:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 94 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OGHzjyl0433cqDu3cFV:-OGHzjyl0433cqDu3cFW:b58xobe) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd29e35e9f2ae3f7a69b2a1b0a94768ad532b589/pkg/sql/vecindex/vecstore/persistent_txn.go#L94)):*
> ```Go
> // encodeVectorFromSet encodes the vector indicated by 'idx' from an external
> // vector set.
> func (psc *persistentStoreCodec) encodeVectorFromSet(
> ```

nit: for this and `Encode*Vector`, it doesn't look like we're ever appending to a non-empty slice, so we could lose the `appendTo` argument. I'm fine leaving that as future refactoring, though.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-:-OGVKNq97UA5FV42jATB:bg7igkv)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 26 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OGVJNov65Pxj79ZSoQs:-OGVJNov65Pxj79ZSoQt:bb36vp9) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd29e35e9f2ae3f7a69b2a1b0a94768ad532b589/pkg/sql/vecindex/vecstore/persistent_txn.go#L26)):*
> ```Go
> // vector index's internal data. Committing changes is the responsibility of the
> // caller.
> type PersistentStoreTxn struct {
> ```

NIT: Why export this type?
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 380 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OGVKEpW2CogFTeJ-8Lz:-OGVKEpW2CogFTeJ-8M-:b-kqbe4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd29e35e9f2ae3f7a69b2a1b0a94768ad532b589/pkg/sql/vecindex/vecstore/persistent_txn.go#L380)):*
> ```Go
> 	}
> 
> 	if err := psTxn.kv.Run(ctx, b); err != nil {
> ```

Does this code path try to parallelize the scans?


<!-- Sent from Reviewable.io -->

mw5h reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137319)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @andy-kimball and @DrewKimball)

___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 236 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFO50vo7iSkaaybeZAI:-OGWgWsT9nIy7-3WIDph:b5tpf53) ([raw file](https://github.com/cockroachdb/cockroach/blob/558ba4c35c2bc62c3b8bb2ebc8d1023b14b8e4d6/pkg/sql/vecindex/vecstore/persistent_txn.go#L236)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I wouldn't think CPut overhead should matter for partition creation, but if it's a concern we could just make it a documented requirement of the function without actually validating it during execution. It's minor though, so I'm fine with whatever you end up deciding.
</blockquote></details>

I'm going to leave this as is for now. Happy to revisit later if we want.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 26 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OGVJNov65Pxj79ZSoQs:-OGWex6C6MUgpNu4mkc7:b-sctdba) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd29e35e9f2ae3f7a69b2a1b0a94768ad532b589/pkg/sql/vecindex/vecstore/persistent_txn.go#L26)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: Why export this type?
</blockquote></details>

It shouldn't need to be exported. Thanks for pointing that out.
___
*[`pkg/sql/vecindex/vecstore/persistent_txn.go` line 380 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OGVKEpW2CogFTeJ-8Lz:-OGWfYYs23zRf821YoCE:b-gl9ycz) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd29e35e9f2ae3f7a69b2a1b0a94768ad532b589/pkg/sql/vecindex/vecstore/persistent_txn.go#L380)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Does this code path try to parallelize the scans?
</blockquote></details>

My understanding is that the KV layer will attempt to parallelize requests received in a batch.
___
*[`pkg/sql/vecindex/vecstore/store_test.go` line 17 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFOEuVf1b3InVKoHlla:-OGWgTnE60WwydxKw2Y4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aad54a088728cbdadf9a29def703bdda000158d/pkg/sql/vecindex/vecstore/store_test.go#L17)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Nice!
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/quantize/quantizer.go` line 120 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/137319#-OFO1cvT9-se3fZmoiEv:-OGWgUiB4WglbS-o-lxJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1f431f952c7b0b52e670d6dd9252c4cb1d5bfc3/pkg/sql/vecindex/quantize/quantizer.go#L120)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I see, too bad.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mw5h commented:
TFTR

mw5h commented:
bors r+

mw5h commented:
bors retry
